swift/mixed_language_library.bzl

```
@@ -0,0 +1,616 @@
+        headers = outputs_depset,
+        direct_public_headers = outputs,
+        includes = depset(includes),
+    )
+
+    return [
+        DefaultInfo(files = outputs_depset),
+        CcInfo(compilation_context = compilation_context),
+    ]
+
+_mixed_language_library_umbrella_header = rule(
+    attrs = {
+        "actual_name": attr.string(mandatory = True),
+        "includes": attr.string_list(),
+        "hdrs": attr.label_list(allow_files = True),
+        "module_name": attr.string(mandatory = True),
+    },
+    doc = "Creates an umbrella header for a mixed language target.",
+    implementation = _mixed_language_library_umbrella_header_impl,
+)
+
+# `_mixed_language_library_internal_module_map`
+
+def _mixed_language_library_internal_module_map_impl(ctx):
+    actions = ctx.actions
+
+    module_map = actions.declare_file(ctx.attr.actual_name + ".modulemap")
+
+    # TODO: Set `dependent_module_names`
+    write_module_map(
+        actions = actions,
+        exported_module_ids = ["*"],
+        module_map_file = module_map,
+        module_name = ctx.attr.module_name,
+        umbrella_header = ctx.file.umbrella_header,
+        public_headers = ctx.files.hdrs,
+        public_textual_headers = ctx.files.textual_hdrs,
+    )
+
+    return [DefaultInfo(files = depset([module_map]))]
+
+_mixed_language_library_internal_module_map = rule(
+    attrs = {
+        "actual_name": attr.string(mandatory = True),
+        "hdrs": attr.label_list(allow_files = True),
+        "module_name": attr.string(mandatory = True),
+        "textual_hdrs": attr.label_list(allow_files = True),
+        "umbrella_header": attr.label(allow_single_file = True),
+    },
+    doc = """\
+Creates the module map file that is used internally by the clang and swift
+library targets in the `mixed_language_library` macro.
+""",
+    implementation = _mixed_language_library_internal_module_map_impl,
+)
+
+# `_mixed_language_library`
+
+def _write_extended_module_map(
+        *,
+        actions,
+        module_name,
+        original_module_map,
+        output_module_map,
+        swift_generated_header):
+    args = actions.args()
+    args.add(output_module_map)
+    args.add(original_module_map)
+    args.add(module_name)
+    args.add(swift_generated_header.basename)
+
+    actions.run_shell(
+        arguments = [args],
+        command = """\
+set -eu
+
+output="$1"
+input="$2"
+module_name="$3"
+generated_header="$4"
+
+# FIXME: Actually parse the map and fix the paths relative to the original
+# location, since it might not be the generated one. We can't use VFS overlay
+# like SE-0403, so rewriting the modulemap is the only way.
+sed 's|header "|header "../../|g' "$input" > "$output"
+
+cat <<EOF >> "$output"
+
+module "$module_name".Swift {
+    header "$generated_header"
+
+    requires objc
+}
+EOF
+""",
+        inputs = [original_module_map],
+        mnemonic = "ExtendMixedLanguageModuleMap",
```

If we move this to `module_map.bzl`:

```suggestion
        mnemonic = "ExtendedModuleMap",
```